lib/moonshot/command.rb

```diff
@@ -0,0 +1,67 @@
+require 'thor'
+
+module Moonshot
+  # A Command that is automatically registered with the Moonshot::CommandLine
+  class Command
+    module ClassMethods
+      # TODO: Can we auto-generate usage for commands with no positional arguments, at least?
+      attr_accessor :usage, :description
+    end
+
+    def self.inherited(base)
+      Moonshot::CommandLine.register(base)
+      base.extend(ClassMethods)
+    end
+
+    def parser
+      @use_interactive_logger = true
+
+      OptionParser.new do |o|
+        o.banner = "Usage: moonshot #{self.class.usage}"
+
+        o.on('-v', '--[no-]verbose', 'Show debug logging') do |v|
+          Moonshot.config.interactive_logger.debug = true if v
+        end
+
+        o.on('-nNAME', '--environment=NAME', 'Which environment to operate on.') do |v|
+          Moonshot.config.environment_name = v
+        end
+
+        o.on('--[no-]interactive-logger', TrueClass, 'Enable or disable fancy logging') do |v|
+          @use_interactive_logger = v
+        end
+
+        # TODO: This is going to be shown for incompatible options, we
+        # could create more class hierarchy to support it?
```

Comment:
A class hierarchy is probably fine, could also be a method (add_stack_options(parser)) or module if you want to avoid hierarchy.

I would suggest the ones at the top level be considered global and can be printed in the top level help.